open-machine-learning-jupyter-book/ml-advanced/clustering/k-means-clustering.md

```diff
@@ -0,0 +1,280 @@
+
+```{code-cell}
+from sklearn.preprocessing import LabelEncoder
+le = LabelEncoder()
+
+X = df.loc[:, ('artist_top_genre','popularity','danceability','acousticness','loudness','energy')]
+
+y = df['artist_top_genre']
+
+X['artist_top_genre'] = le.fit_transform(X['artist_top_genre'])
+
+y = le.transform(y)
+```
+
+3. Now you need to pick how many clusters to target. You know there are 3 song genres that we carved out of the dataset, so let's try 3:
+
+```{code-cell}
+from sklearn.cluster import KMeans
+
+nclusters = 3 
+seed = 0
+
+km = KMeans(n_clusters=nclusters, random_state=seed)
+km.fit(X)
+
+# Predict the cluster for each data point
+
+y_cluster_kmeans = km.predict(X)
+y_cluster_kmeans
+```
+
+You see an array printed out with predicted clusters (0, 1,or 2) for each row of the dataframe.
+
+4. Use this array to calculate a 'silhouette score':
+
+```{code-cell}
+from sklearn import metrics
+score = metrics.silhouette_score(X, y_cluster_kmeans)
+score
+```
+
+## Silhouette score
+
+Look for a silhouette score closer to 1. This score varies from -1 to 1, and if the score is 1, the cluster is dense and well-separated from other clusters. A value near 0 represents overlapping clusters with samples very close to the decision boundary of the neighboring clusters.
+
+Our score is **.53**, so right in the middle. This indicates that our data is not particularly well-suited to this type of clustering, but let's continue.
+
+### Exercise - build a model
+
+1. Import `KMeans` and start the clustering process.
+
+```{code-cell}
+from sklearn.cluster import KMeans
+wcss = []
+
+for i in range(1, 11):
+    kmeans = KMeans(n_clusters = i, init = 'k-means++', random_state = 42)
+    kmeans.fit(X)
+    wcss.append(kmeans.inertia_)
+```
+
+ There are a few parts here that warrant explaining.
+
+```{seealso} 
+🎓 range: These are the iterations of the clustering process
+
+🎓 random_state: "Determines random number generation for centroid initialization."
+
+🎓 WCSS: "within-cluster sums of squares" measures the squared average distance of all the points within a cluster to the cluster centroid.
+
+🎓 Inertia: K-Means algorithms attempt to choose centroids to minimize 'inertia', "a measure of how internally coherent clusters are." The value is appended to the wcss variable on each iteration.
+
+🎓 k-means++: In Scikit-learn you can use the 'k-means++' optimization, which "initializes the centroids to be (generally) distant from each other, leading to probably better results than random initialization.
+```
+
+### Elbow method
+
+Previously, you surmised that, because you have targeted 3 song genres, you should choose 3 clusters. But is that the case?
+
+1. Use the 'elbow method' to make sure.
+
+```sql
+plt.figure(figsize=(10,5))
+sns.lineplot(range(1, 11), wcss,marker='o',color='red')
+plt.title('Elbow')
+plt.xlabel('Number of clusters')
+plt.ylabel('WCSS')
+plt.show()
+```
+
+```{figure} ../../../images/clustering/elbow.png
+---
+name: elbow
+---
+```
```

Comment:
If above becomes code-cell, the image will be generated and you don't need this anymore.